aggregator/handler/sharded.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package handler
+
+import (
+	"github.com/m3db/m3aggregator/aggregator"
+	"github.com/m3db/m3aggregator/aggregator/handler/common"
+	"github.com/m3db/m3aggregator/aggregator/handler/writer"
+	"github.com/m3db/m3aggregator/sharding"
+
+	"github.com/uber-go/tally"
+)
+
+// SharderRouter contains a sharder id and a router.
+type SharderRouter struct {
+	sharding.SharderID
+	common.Router
+}
+
+type sharderRouters struct {
+	SharderID sharding.SharderID
+	Routers   []common.Router
+}
+
+type shardedHandler struct {
+	routersBySharderID []SharderRouter
+	writerOpts         writer.Options
+}
+
+func newShardedHandler(srs []SharderRouter, writerOpts writer.Options) aggregator.Handler {
+	// Group routers by their sharder ids so that metrics are only encoded once
+	// for backends with the same sharding functions.
+	var (
+		nonShardedRouters  = make([]common.Router, 0, len(srs))
+		shardedRoutersByID = make([]sharderRouters, 0, len(srs))
+	)
+	for _, sr := range srs {
+		if sr.SharderID == sharding.NoShardingSharderID {
+			nonShardedRouters = append(nonShardedRouters, sr.Router)
+		} else {
+			found := false
+			for i := range shardedRoutersByID {
+				if shardedRoutersByID[i].SharderID == sr.SharderID {
+					shardedRoutersByID[i].Routers = append(shardedRoutersByID[i].Routers, sr.Router)
+					found = true
+					break
+				}
+			}
+			if !found {
+				shardedRoutersByID = append(shardedRoutersByID, sharderRouters{
+					SharderID: sr.SharderID,
+					Routers:   []common.Router{sr.Router},
+				})
+			}
+		}
+	}
+
+	// If there are both sharded routers and non-sharded routers, the non-sharded
```

Comment:
nit: do you think we'll have both non-sharded and sharded backends a lot in practice? This feels like a premature optimizations perhaps.

Reply:
We have both sharded and non-sharded backends (ingester+indexer). Without this optimization we are paying the encoding the cost doubly for the non-sharded backends, which is on the critical path of flushes and can definitely negatively affect the performance if the flush latency gets out of control.

Given that the optimization only adds one extra line, I'd be willing to do that for the benefit of halving the cycles spent on encoding. This also enables users in the OSS world to add both sharded and non-sharded backends with relatively low cost.